File: src/interface/UdsClientManager.ts
/**
 * UDS客户端管理器 - 用于React应用
 * 提供高级的UDS诊断服务接口
 */

import { DoipClient, DoipClientConfig } from './DoipClient';
import { UdsService, UdsConfig, UdsResponse } from './UdsService';

export interface DiagnosticResult {
  success: boolean;
  message: string;
  data?: any;
  timestamp: string;
}

export interface ConnectionConfig {
  ipAddress: string;
  port: number;
  serverAddress: string;
  clientAddress: string;
  timeout?: number;
}

export class UdsClientManager {
  private doipClient: DoipClient | null = null;
  private udsService: UdsService | null = null;
  private isConnected: boolean = false;
  private connectionConfig: ConnectionConfig | null = null;

  /**
   * 连接到ECU
   */
  async connect(config: ConnectionConfig): Promise<DiagnosticResult> {
    try {
      this.connectionConfig = config;

      const doipConfig: DoipClientConfig = {
        ipAddress: config.ipAddress,
        port: config.port,
        timeout: config.timeout || 30000
      };

      this.doipClient = new DoipClient(doipConfig);

      const connected = await this.doipClient.connect();
      if (!connected) {
        return {
          success: false,
          message: '连接ECU失败',
          timestamp: new Date().toISOString()
        };
      }

      // 创建UDS服务
      const udsConfig: UdsConfig = {
        vehicleInfo: {
          serverAddress: config.serverAddress,
          clientAddress: config.clientAddress
        }
      };

      this.udsService = new UdsService(this.doipClient, udsConfig);

      // 执行路由激活
      const routineActive = await this.udsService.routineActive();
      if (!routineActive) {
        await this.disconnect();
        return {
          success: false,
          message: '路由激活失败',
          timestamp: new Date().toISOString()
        };
      }

      this.isConnected = true;
      return {
        success: true,
        message: `成功连接到ECU ${config.ipAddress}:${config.port}`,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        message: `连接失败: ${error}`,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * 断开连接
   */
  async disconnect(): Promise<DiagnosticResult> {
    try {
      if (this.doipClient) {
        await this.doipClient.disconnect();
        this.doipClient = null;
      }

      this.udsService = null;
      this.isConnected = false;

      return {
        success: true,
        message: '已断开ECU连接',
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        message: `断开连接失败: ${error}`,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * 检查连接状态
   */
  getConnectionStatus(): boolean {
    return this.isConnected && this.doipClient?.isSocketConnected() === true;
  }

  /**
   * 发送UDS命令
   */
  async sendUdsCommand(serviceId: string, data: string): Promise<DiagnosticResult> {
    if (!this.udsService || !this.isConnected) {
      return {
        success: false,
        message: '未连接到ECU',
        timestamp: new Date().toISOString()
      };
    }

    try {
      const service = parseInt(serviceId.replace('0x', ''), 16);
      const dataBytes = this.hexStringToBytes(data);

      let result: boolean | UdsResponse;
      let message: string;

      switch (service) {
        case 0x10: // 诊断会话控制
          const session = dataBytes.length > 1 ? dataBytes[1] : 0x01;
          result = await this.udsService.startSession(session);
          message = result ? '会话控制成功' : '会话控制失败';
          break;

        case 0x11: // ECU重启
          const resetType = dataBytes.length > 1 ? dataBytes[1] : 0x01;
          result = await this.udsService.ecuReset(resetType);
          message = result ? 'ECU重启成功' : 'ECU重启失败';
          break;

        case 0x14: // 清除DTC
          result = await this.udsService.clearDiagnosticInformation();
          message = result ? '清除DTC成功' : '清除DTC失败';
          break;

        case 0x19: // 读取DTC信息
          const subFunc = dataBytes.length > 1 ? dataBytes[1] : 0x02;
          result = await this.udsService.readDtcInformation(subFunc);
          message = result.success ? '读取DTC信息成功' : '读取DTC信息失败';
          break;

        case 0x22: // 读取数据标识符
          if (dataBytes.length >= 3) {
            const did = (dataBytes[1] << 8) | dataBytes[2];
            result = await this.udsService.readDataByIdentifier(did);
            message = result.success ? `读取DID 0x${did.toString(16)}成功` : `读取DID 0x${did.toString(16)}失败`;
          } else {
            throw new Error('DID参数不足');
          }
          break;

        case 0x27: // 安全访问
          if (dataBytes.length >= 2) {
            const level = dataBytes[1];
            if (level % 2 === 1) {
              // 奇数级别：获取种子
              result = await this.udsService.securityAccessGetSeed(level);
              message = result ? '获取安全访问种子成功' : '获取安全访问种子失败';
            } else {
              // 偶数级别：发送密钥（这里需要实际的密钥计算）
              const key = 0x1234; // 示例密钥，实际应用中需要正确计算
              result = await this.udsService.securityAccessCompareKey(level, key);
              message = result ? '安全访问验证成功' : '安全访问验证失败';
            }
          } else {
            throw new Error('安全访问级别参数不足');
          }
          break;

        case 0x28: // 通信控制
          const commType = dataBytes.length > 1 ? dataBytes[1] : 0x00;
          result = await this.udsService.communicationControl(commType);
          message = result ? '通信控制成功' : '通信控制失败';
          break;

        case 0x2E: // 写入数据标识符
          if (dataBytes.length >= 3) {
            const did = (dataBytes[1] << 8) | dataBytes[2];
            const writeData = new TextDecoder().decode(dataBytes.slice(3));
            result = await this.udsService.writeDataByIdentifier(did, writeData);
            message = result.success ? `写入DID 0x${did.toString(16)}成功` : `写入DID 0x${did.toString(16)}失败`;
          } else {
            throw new Error('写入DID参数不足');
          }
          break;

        case 0x3E: // 测试器在线
          const suppressResp = dataBytes.length > 1 && dataBytes[1] === 0x80;
          result = await this.udsService.testerPresent(suppressResp);
          message = result ? '测试器在线成功' : '测试器在线失败';
          break;

        case 0x85: // 控制DTC设置
          const dtcType = dataBytes.length > 1 ? dataBytes[1] : 0x02;
          result = await this.udsService.controlDtcSetting(dtcType);
          message = result ? 'DTC控制成功' : 'DTC控制失败';
          break;

        default:
          throw new Error(`不支持的UDS服务: 0x${service.toString(16)}`);
      }

      const success = typeof result === 'boolean' ? result : result.success;
      const responseData = typeof result === 'object' ? result.data : undefined;

      return {
        success,
        message,
        data: responseData,
        timestamp: new Date().toISOString()
      };
    } catch (error) {
      return {
        success: false,
        message: `UDS命令执行失败: ${error}`,
        timestamp: new Date().toISOString()
      };
    }
  }

  /**
   * 获取连接配置
   */
  getConnectionConfig(): ConnectionConfig | null {
    return this.connectionConfig;
  }

  /**
   * 工具函数：十六进制字符串转字节数组
   */
  private hexStringToBytes(hex: string): Uint8Array {
    const cleanHex = hex.replace(/\s+/g, '');
    const bytes = new Uint8Array(cleanHex.length / 2);
    for (let i = 0; i < cleanHex.length; i += 2) {
      bytes[i / 2] = parseInt(cleanHex.substring(i, i + 2), 16);
    }
    return bytes;
  }
}

// 创建全局实例
export const udsClientManager = new UdsClientManager();

File: src-tauri/src/uds_client_manager.rs
/**
 * UDS 客户端管理器 - Rust 实现
 * 提供高级的 UDS 诊断服务接口，用于 Tauri 应用
 */
use crate::doip_client::DoipClient;
use crate::types::{ConnectionConfig, DiagnosticResult, DoipClientConfig, UdsConfig, UdsServices};
use crate::uds_service::UdsService;
use crate::utils::{get_timestamp, hex_to_bytes};

pub struct UdsClientManager {
    uds_service: Option<UdsService>,
    is_connected: bool,
    connection_config: Option<ConnectionConfig>,
}

impl UdsClientManager {
    /// 创建新的 UDS 客户端管理器
    pub fn new() -> Self {
        Self {
            uds_service: None,
            is_connected: false,
            connection_config: None,
        }
    }

    /// 连接到 ECU
    pub async fn connect(&mut self, config: ConnectionConfig) -> DiagnosticResult {
        self.connection_config = Some(config.clone());

        let doip_config = DoipClientConfig {
            ip_address: config.ip_address.clone(),
            port: config.port,
            timeout: config.timeout,
        };

        let mut doip_client = DoipClient::new(doip_config);

        match doip_client.connect().await {
            Ok(true) => {
                println!("TCP connection established successfully");

                // 创建 UDS 服务
                let uds_config = UdsConfig {
                    vehicle_info: crate::types::VehicleConfig {
                        server_address: config.server_address.clone(),
                        client_address: config.client_address.clone(),
                    },
                };

                println!(
                    "Creating UDS service with server_address: {}, client_address: {}",
                    config.server_address, config.client_address
                );

                match UdsService::new(doip_client, uds_config) {
                    Ok(mut uds_service) => {
                        // 执行路由激活
                        match uds_service.routine_active().await {
                            Ok(true) => {
                                self.uds_service = Some(uds_service);
                                self.is_connected = true;

                                DiagnosticResult {
                                    success: true,
                                    message: format!(
                                        "成功连接到ECU {}:{} 并完成路由激活",
                                        config.ip_address, config.port
                                    ),
                                    data: None,
                                    timestamp: get_timestamp(),
                                }
                            }
                            Ok(false) => DiagnosticResult {
                                success: false,
                                message: "路由激活被拒绝".to_string(),
                                data: None,
                                timestamp: get_timestamp(),
                            },
                            Err(e) => {
                                // 记录详细的路由激活失败信息
                                eprintln!("Routing activation failed: {:?}", e);
                                DiagnosticResult {
                                    success: false,
                                    message: format!("路由激活失败: {:?}", e),
                                    data: None,
                                    timestamp: get_timestamp(),
                                }
                            }
                        }
                    }
                    Err(e) => DiagnosticResult {
                        success: false,
                        message: format!("创建UDS服务失败: {}", e),
                        data: None,
                        timestamp: get_timestamp(),
                    },
                }
            }
            Ok(false) => DiagnosticResult {
                success: false,
                message: "连接ECU失败".to_string(),
                data: None,
                timestamp: get_timestamp(),
            },
            Err(e) => DiagnosticResult {
                success: false,
                message: format!("连接ECU失败: {:?}", e),
                data: None,
                timestamp: get_timestamp(),
            },
        }
    }

    /// 断开连接
    pub async fn disconnect(&mut self) -> DiagnosticResult {
        self.uds_service = None;
        self.is_connected = false;

        DiagnosticResult {
            success: true,
            message: "已断开ECU连接".to_string(),
            data: None,
            timestamp: get_timestamp(),
        }
    }

    /// 检查连接状态
    pub fn get_connection_status(&self) -> bool {
        self.is_connected && self.uds_service.is_some()
    }

    /// 发送 UDS 命令
    pub async fn send_uds_command(&mut self, service_id: &str, data: &str) -> DiagnosticResult {
        if !self.is_connected || self.uds_service.is_none() {
            return DiagnosticResult {
                success: false,
                message: "未连接到ECU".to_string(),
                data: None,
                timestamp: get_timestamp(),
            };
        }

        let service = match u8::from_str_radix(service_id.trim_start_matches("0x"), 16) {
            Ok(s) => s,
            Err(_) => {
                return DiagnosticResult {
                    success: false,
                    message: format!("无效的服务ID: {}", service_id),
                    data: None,
                    timestamp: get_timestamp(),
                };
            }
        };

        let data_bytes = match self.hex_string_to_bytes(data) {
            Ok(bytes) => bytes,
            Err(e) => {
                return DiagnosticResult {
                    success: false,
                    message: format!("无效的数据格式: {}", e),
                    data: None,
                    timestamp: get_timestamp(),
                };
            }
        };

        let uds_service = self.uds_service.as_mut().unwrap();

        let result = match service {
            UdsServices::DIAGNOSTIC_SESSION_CONTROL => {
                let session = if data_bytes.len() > 1 {
                    data_bytes[1]
                } else {
                    0x01
                };
                match uds_service.start_session(session).await {
                    Ok(success) => (success, "会话控制".to_string(), None),
                    Err(e) => (false, format!("会话控制失败: {}", e), None),
                }
            }

            UdsServices::ECU_RESET => {
                let reset_type = if data_bytes.len() > 1 {
                    data_bytes[1]
                } else {
                    0x01
                };
                match uds_service.ecu_reset(reset_type).await {
                    Ok(success) => (success, "ECU重启".to_string(), None),
                    Err(e) => (false, format!("ECU重启失败: {}", e), None),
                }
            }

            UdsServices::CLEAR_DIAGNOSTIC_INFORMATION => {
                match uds_service.clear_diagnostic_information().await {
                    Ok(success) => (success, "清除DTC".to_string(), None),
                    Err(e) => (false, format!("清除DTC失败: {}", e), None),
                }
            }

            UdsServices::READ_DTC_INFORMATION => {
                let sub_func = if data_bytes.len() > 1 {
                    data_bytes[1]
                } else {
                    0x02
                };
                match uds_service.read_dtc_information(sub_func).await {
                    Ok(response) => (response.success, "读取DTC信息".to_string(), response.data),
                    Err(e) => (false, format!("读取DTC信息失败: {}", e), None),
                }
            }

            UdsServices::READ_DATA_BY_IDENTIFIER => {
                if data_bytes.len() >= 3 {
                    let did = ((data_bytes[1] as u16) << 8) | (data_bytes[2] as u16);
                    match uds_service.read_data_by_identifier(did).await {
                        Ok(response) => (
                            response.success,
                            format!("读取DID 0x{:04x}", did),
                            response.data,
                        ),
                        Err(e) => (false, format!("读取DID失败: {}", e), None),
                    }
                } else {
                    (false, "DID参数不足".to_string(), None)
                }
            }

            UdsServices::SECURITY_ACCESS => {
                if data_bytes.len() >= 2 {
                    let level = data_bytes[1];
                    if level % 2 == 1 {
                        // 奇数级别：获取种子
                        match uds_service.security_access_get_seed(level).await {
                            Ok(success) => (success, "获取安全访问种子".to_string(), None),
                            Err(e) => (false, format!("获取安全访问种子失败: {}", e), None),
                        }
                    } else {
                        // 偶数级别：发送密钥（这里需要实际的密钥计算）
                        let key = 0x1234u32; // 示例密钥，实际应用中需要正确计算
                        match uds_service.security_access_compare_key(level, key).await {
                            Ok(success) => (success, "安全访问验证".to_string(), None),
                            Err(e) => (false, format!("安全访问验证失败: {}", e), None),
                        }
                    }
                } else {
                    (false, "安全访问级别参数不足".to_string(), None)
                }
            }

            UdsServices::COMMUNICATION_CONTROL => {
                let comm_type = if data_bytes.len() > 1 {
                    data_bytes[1]
                } else {
                    0x00
                };
                match uds_service.communication_control(comm_type).await {
                    Ok(success) => (success, "通信控制".to_string(), None),
                    Err(e) => (false, format!("通信控制失败: {}", e), None),
                }
            }

            UdsServices::WRITE_DATA_BY_IDENTIFIER => {
                if data_bytes.len() >= 3 {
                    let did = ((data_bytes[1] as u16) << 8) | (data_bytes[2] as u16);
                    let write_data = String::from_utf8_lossy(&data_bytes[3..]);
                    match uds_service.write_data_by_identifier(did, &write_data).await {
                        Ok(response) => (
                            response.success,
                            format!("写入DID 0x{:04x}", did),
                            response.data,
                        ),
                        Err(e) => (false, format!("写入DID失败: {}", e), None),
                    }
                } else {
                    (false, "写入DID参数不足".to_string(), None)
                }
            }

            UdsServices::TESTER_PRESENT => {
                let suppress_resp = data_bytes.len() > 1 && data_bytes[1] == 0x80;
                match uds_service.tester_present(suppress_resp).await {
                    Ok(success) => (success, "测试器在线".to_string(), None),
                    Err(e) => (false, format!("测试器在线失败: {}", e), None),
                }
            }

            UdsServices::CONTROL_DTC_SETTING => {
                let dtc_type = if data_bytes.len() > 1 {
                    data_bytes[1]
                } else {
                    0x02
                };
                match uds_service.control_dtc_setting(dtc_type).await {
                    Ok(success) => (success, "DTC控制".to_string(), None),
                    Err(e) => (false, format!("DTC控制失败: {}", e), None),
                }
            }

            _ => (false, format!("不支持的UDS服务: 0x{:02x}", service), None),
        };

        let (success, message, response_data) = result;
        let final_message = if success {
            format!("{}成功", message)
        } else {
            message
        };

        DiagnosticResult {
            success,
            message: final_message,
            data: response_data.map(|d| serde_json::Value::String(hex::encode(d))),
            timestamp: get_timestamp(),
        }
    }

    /// 获取连接配置
    pub fn get_connection_config(&self) -> Option<&ConnectionConfig> {
        self.connection_config.as_ref()
    }

    /// 工具函数：十六进制字符串转字节数组
    fn hex_string_to_bytes(&self, hex: &str) -> Result<Vec<u8>, String> {
        hex_to_bytes(hex).map_err(|e| e.to_string())
    }
}

impl Default for UdsClientManager {
    fn default() -> Self {
        Self::new()
    }
}
